sklearn/naive_bayes.py

```diff
@@ -726,6 +726,111 @@ def _joint_log_likelihood(self, X):
+    assumptions" made by the standard Multinomial Naive Bayes classifier. See
+    Rennie et al. (2003) for further discussion.
+
+    Parameters
+    ----------
+    alpha : float, optional (default=1.0)
+        Additive (Laplace/Lidstone) smoothing parameter
+        (0 for no smoothing).
+
+    fit_prior : boolean, optional (default=False)
+        Whether to learn class prior probabilities or not.
+        If false, a uniform prior will be used.
+
+    class_prior : array-like, size (n_classes,), optional (default=None)
+        Prior probabilities of the classes. If specified the priors are not
+        adjusted according to the data.
+
+    Attributes
+    ----------
+    class_log_prior_ : array, shape (n_classes, )
+        Smoothed empirical log probability for each class.
+
+    intercept_ : property
+        Mirrors ``class_log_prior_`` for interpreting MultinomialNB
+        as a linear model.
+
+    feature_log_prob_ : array, shape (n_classes, n_features)
+        Empirical log probability of features
+        given a class, ``P(x_i|y)``.
+
+    coef_ : property
+        Mirrors ``feature_log_prob_`` for interpreting MultinomialNB
+        as a linear model.
+
+    class_count_ : array, shape (n_classes,)
+        Number of samples encountered for each class during fitting. This
+        value is weighted by the sample weight when provided.
+
+    feature_count_ : array, shape (n_classes, n_features)
+        Number of samples encountered for each (class, feature)
+        during fitting. This value is weighted by the sample weight when
+        provided.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> X = np.random.randint(5, size=(6, 100))
+    >>> y = np.array([1, 2, 3, 4, 5, 6])
+    >>> from sklearn.naive_bayes import MultinomialCNB
+    >>> clf = MultinomialCNB()
+    >>> clf.fit(X, y)
+    MultinomialCNB(alpha=1.0, class_prior=None, fit_prior=False)
+    >>> print(clf.predict(X[2:3]))
+    [3]
+
+    References
+    ----------
+    Rennie, J. D., Shih, L., Teevan, J., & Karger, D. R. (2003, August).
+    Tackling the poor assumptions of naive bayes text classifiers. In ICML
+    (Vol. 3, pp. 616-623).
+    http://people.csail.mit.edu/jrennie/papers/icml03-nb.pdf
+    """
+
+    def __init__(self, alpha=1.0, fit_prior=False, class_prior=None):
+        self.alpha = alpha
+        self.fit_prior = fit_prior
+        self.class_prior = class_prior
+
+    def _count(self, X, Y):
+        """Count and smooth feature occurrences and compute weights."""
+        if np.any((X.data if issparse(X) else X) < 0):
+            raise ValueError("Input X must be non-negative")
+        self.feature_count_ += safe_sparse_dot(Y.T, X)
+        self.class_count_ += Y.sum(axis=0)
+
+        y = Y.argmax(1)
+        n_classes = len(self.classes_)
+        n_features = X.shape[1]
+        weights = np.zeros((n_classes, n_features), dtype=np.float64)
+        numerator_all = X.sum(axis=0) + self.alpha
```

Comment:
The code is hopefully a little clearer now. I'm implementing steps four through eight of the algorithm outlined in Table 4 of the paper. Basically, you count up the features (with a smoothing factor) for the complement of each class, normalize those counts, take the logarithm of these new weights, and then normalize again. The input matrix should just be non-negative (e.g., it can be either a tf-idf matrix or a raw term count matrix).